Exception when purging users through the UI

> However, a separate issue now occurs when trying to purge users through the UI, this is on 24.2: 
> 
> ```
> Aug 21 16:57:02: uvicorn.access INFO 2025-08-21 16:57:02,465 [pN:main.4,p:3338777,tN:MainThread] 94.105.110.46:0 - "GET /history/current_history_json?since=2025-08-20T14:02:28.828474 HTTP/1.0" 200
> Aug 21 16:57:04: uvicorn.access INFO 2025-08-21 16:57:04,757 [pN:main.2,p:3341198,tN:MainThread] 193.190.80.1:0 - "GET /api/jobs/8aa6514a3d989754 HTTP/1.0" 200
> Aug 21 16:57:05: galaxy.webapps.galaxy.api.users DEBUG 2025-08-21 16:57:05,909 [pN:main.4,p:3338777,tN:AnyIO worker thread] Purging user <galaxy.model.User(17) at 0x7f39b28496d0>
> Aug 21 16:57:05: uvicorn.access INFO 2025-08-21 16:57:05,921 [pN:main.2,p:3341198,tN:MainThread] 94.105.110.46:0 - "GET /history/current_history_json?since=2025-08-20T14:02:28.828474 HTTP/1.0" 200
> Aug 21 16:57:05: COMPLIANCE INFO 2025-08-21 16:57:05,933 [pN:main.4,p:3338777,tN:AnyIO worker thread] delete-user-event: gdprtest2
> Aug 21 16:57:05: uvicorn.access INFO 2025-08-21 16:57:05,935 [pN:main.4,p:3338777,tN:MainThread] 94.105.110.46:0 - "DELETE /api/users/7d12165cdc09ed44?purge=true HTTP/1.0" 500
> Aug 21 16:57:05: [2025-08-21 16:57:05 +0200] [3338777] [ERROR] Exception in ASGI application
> Aug 21 16:57:05:   + Exception Group Traceback (most recent call last):
> Aug 21 16:57:05:   |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_utils.py", line 76, in collapse_excgroups
> Aug 21 16:57:05:   |     yield
> Aug 21 16:57:05:   |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 186, in __call__
> Aug 21 16:57:05:   |     async with anyio.create_task_group() as task_group:
> Aug 21 16:57:05:   |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 685, in __aexit__
> Aug 21 16:57:05:   |     raise BaseExceptionGroup(
> Aug 21 16:57:05:   | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
> Aug 21 16:57:05:   +-+---------------- 1 ----------------
> Aug 21 16:57:05:     | Traceback (most recent call last):
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/uvicorn/protocols/http/h11_impl.py", line 406, in run_asgi
> Aug 21 16:57:05:     |     result = await app(  # type: ignore[func-returns-value]
> Aug 21 16:57:05:     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
> Aug 21 16:57:05:     |     return await self.app(scope, receive, send)
> Aug 21 16:57:05:     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
> Aug 21 16:57:05:     |     await super().__call__(scope, receive, send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/applications.py", line 113, in __call__
> Aug 21 16:57:05:     |     await self.middleware_stack(scope, receive, send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 187, in __call__
> Aug 21 16:57:05:     |     raise exc
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 165, in __call__
> Aug 21 16:57:05:     |     await self.app(scope, receive, _send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette_context/middleware/raw_middleware.py", line 92, in __call__
> Aug 21 16:57:05:     |     await self.app(scope, receive, send_wrapper)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 185, in __call__
> Aug 21 16:57:05:     |     with collapse_excgroups():
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/miniconda3/envs/_galaxy_/lib/python3.11/contextlib.py", line 158, in __exit__
> Aug 21 16:57:05:     |     self.gen.throw(typ, value, traceback)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
> Aug 21 16:57:05:     |     raise exc
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 187, in __call__
> Aug 21 16:57:05:     |     response = await self.dispatch_func(request, call_next)
> Aug 21 16:57:05:     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/server/lib/galaxy/webapps/galaxy/fast_app.py", line 108, in add_x_frame_options
> Aug 21 16:57:05:     |     response = await call_next(request)
> Aug 21 16:57:05:     |                ^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 163, in call_next
> Aug 21 16:57:05:     |     raise app_exc
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 149, in coro
> Aug 21 16:57:05:     |     await self.app(scope, receive_or_disconnect, send_no_error)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
> Aug 21 16:57:05:     |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
> Aug 21 16:57:05:     |     raise exc
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
> Aug 21 16:57:05:     |     await app(scope, receive, sender)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 715, in __call__
> Aug 21 16:57:05:     |     await self.middleware_stack(scope, receive, send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 735, in app
> Aug 21 16:57:05:     |     await route.handle(scope, receive, send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 288, in handle
> Aug 21 16:57:05:     |     await self.app(scope, receive, send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 76, in app
> Aug 21 16:57:05:     |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
> Aug 21 16:57:05:     |     raise exc
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
> Aug 21 16:57:05:     |     await app(scope, receive, sender)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 73, in app
> Aug 21 16:57:05:     |     response = await f(request)
> Aug 21 16:57:05:     |                ^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/fastapi/routing.py", line 301, in app
> Aug 21 16:57:05:     |     raw_response = await run_endpoint_function(
> Aug 21 16:57:05:     |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/fastapi/routing.py", line 214, in run_endpoint_function
> Aug 21 16:57:05:     |     return await run_in_threadpool(dependant.call, **values)
> Aug 21 16:57:05:     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool
> Aug 21 16:57:05:     |     return await anyio.to_thread.run_sync(func, *args)
> Aug 21 16:57:05:     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/anyio/to_thread.py", line 56, in run_sync
> Aug 21 16:57:05:     |     return await get_async_backend().run_sync_in_worker_thread(
> Aug 21 16:57:05:     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 2364, in run_sync_in_worker_thread
> Aug 21 16:57:05:     |     return await future
> Aug 21 16:57:05:     |            ^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 864, in run
> Aug 21 16:57:05:     |     result = context.run(func, *args)
> Aug 21 16:57:05:     |              ^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/server/lib/galaxy/webapps/galaxy/api/users.py", line 708, in delete
> Aug 21 16:57:05:     |     self.service.user_manager.purge(user_to_update)
> Aug 21 16:57:05:     |   File "/srv/galaxy_test/shared/server/lib/galaxy/managers/users.py", line 244, in purge
> Aug 21 16:57:05:     |     role.description = role.description.replace(user.username, uname_hash)
> Aug 21 16:57:05:     |                        ^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:     | AttributeError: 'NoneType' object has no attribute 'replace'
> Aug 21 16:57:05:     +------------------------------------
> Aug 21 16:57:05: During handling of the above exception, another exception occurred:
> Aug 21 16:57:05: Traceback (most recent call last):
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/uvicorn/protocols/http/h11_impl.py", line 406, in run_asgi
> Aug 21 16:57:05:     result = await app(  # type: ignore[func-returns-value]
> Aug 21 16:57:05:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
> Aug 21 16:57:05:     return await self.app(scope, receive, send)
> Aug 21 16:57:05:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
> Aug 21 16:57:05:     await super().__call__(scope, receive, send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/applications.py", line 113, in __call__
> Aug 21 16:57:05:     await self.middleware_stack(scope, receive, send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 187, in __call__
> Aug 21 16:57:05:     raise exc
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 165, in __call__
> Aug 21 16:57:05:     await self.app(scope, receive, _send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette_context/middleware/raw_middleware.py", line 92, in __call__
> Aug 21 16:57:05:     await self.app(scope, receive, send_wrapper)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 185, in __call__
> Aug 21 16:57:05:     with collapse_excgroups():
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/miniconda3/envs/_galaxy_/lib/python3.11/contextlib.py", line 158, in __exit__
> Aug 21 16:57:05:     self.gen.throw(typ, value, traceback)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
> Aug 21 16:57:05:     raise exc
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 187, in __call__
> Aug 21 16:57:05:     response = await self.dispatch_func(request, call_next)
> Aug 21 16:57:05:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/server/lib/galaxy/webapps/galaxy/fast_app.py", line 108, in add_x_frame_options
> Aug 21 16:57:05:     response = await call_next(request)
> Aug 21 16:57:05:                ^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 163, in call_next
> Aug 21 16:57:05:     raise app_exc
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 149, in coro
> Aug 21 16:57:05:     await self.app(scope, receive_or_disconnect, send_no_error)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
> Aug 21 16:57:05:     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
> Aug 21 16:57:05:     raise exc
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
> Aug 21 16:57:05:     await app(scope, receive, sender)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 715, in __call__
> Aug 21 16:57:05:     await self.middleware_stack(scope, receive, send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 735, in app
> Aug 21 16:57:05:     await route.handle(scope, receive, send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 288, in handle
> Aug 21 16:57:05:     await self.app(scope, receive, send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 76, in app
> Aug 21 16:57:05:     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
> Aug 21 16:57:05:     raise exc
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
> Aug 21 16:57:05:     await app(scope, receive, sender)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/routing.py", line 73, in app
> Aug 21 16:57:05:     response = await f(request)
> Aug 21 16:57:05:                ^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/fastapi/routing.py", line 301, in app
> Aug 21 16:57:05:     raw_response = await run_endpoint_function(
> Aug 21 16:57:05:                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/fastapi/routing.py", line 214, in run_endpoint_function
> Aug 21 16:57:05:     return await run_in_threadpool(dependant.call, **values)
> Aug 21 16:57:05:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool
> Aug 21 16:57:05:     return await anyio.to_thread.run_sync(func, *args)
> Aug 21 16:57:05:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/anyio/to_thread.py", line 56, in run_sync
> Aug 21 16:57:05:     return await get_async_backend().run_sync_in_worker_thread(
> Aug 21 16:57:05:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 2364, in run_sync_in_worker_thread
> Aug 21 16:57:05:     return await future
> Aug 21 16:57:05:            ^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/database/tool-dependencies/venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 864, in run
> Aug 21 16:57:05:     result = context.run(func, *args)
> Aug 21 16:57:05:              ^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/server/lib/galaxy/webapps/galaxy/api/users.py", line 708, in delete
> Aug 21 16:57:05:     self.service.user_manager.purge(user_to_update)
> Aug 21 16:57:05:   File "/srv/galaxy_test/shared/server/lib/galaxy/managers/users.py", line 244, in purge
> Aug 21 16:57:05:     role.description = role.description.replace(user.username, uname_hash)
> Aug 21 16:57:05:                        ^^^^^^^^^^^^^^^^^^^^^^^^
> Aug 21 16:57:05: AttributeError: 'NoneType' object has no attribute 'replace'
> Aug 21 16:57:05: uvicorn.access INFO 2025-08-21 16:57:05,984 [pN:main.2,p:3338773,tN:MainThread] 94.105.110.46:0 - "GET /api/entry_points?running=true HTTP/1.0" 200
> Aug 21 16:57:06 : uvicorn.access INFO 2025-08-21 16:57:06,022 [pN:main.4,p:3341202,tN:MainThread] 94.105.110.46:0 - "GET /admin/users_list?limit=25&offset=0&search=is%3Adeleted&sort_by=email&sort_desc=true HTTP/1.0" 200
> ```
> 
> Additionally, the username does show up in the log related to the purge event:
> 
> `Aug 21 16:57:05: COMPLIANCE INFO 2025-08-21 16:57:05,933 [pN:main.4,p:3338777,tN:AnyIO worker thread] delete-user-event: gdprtest2` 

 _Originally posted by @pauldg in [#20773](https://github.com/galaxyproject/galaxy/issues/20773#issuecomment-3223929464)_

--- a/lib/galaxy/managers/users.py
+++ b/lib/galaxy/managers/users.py
@@ -241,11 +241,13 @@ class UserManager(base.ModelManager, deletable.PurgableManagerMixin):
         for role in user.all_roles():
             if self.app.config.redact_username_during_deletion:
                 role.name = role.name.replace(user.username, uname_hash)
-                role.description = role.description.replace(user.username, uname_hash)
+                if role.description:
+                    role.description = role.description.replace(user.username, uname_hash)
 
             if self.app.config.redact_email_during_deletion:
                 role.name = role.name.replace(user.email, email_hash)
-                role.description = role.description.replace(user.email, email_hash)
+                if role.description:
+                    role.description = role.description.replace(user.email, email_hash)
             self.session().add(role)
         private_role.name = email_hash
         private_role.description = f"Private Role for {email_hash}"

